File: Day 1/Historian_Hysteria.py
# Part 1
with open("D1.txt", "r") as f:
    data = f.read().splitlines()
    a, b = sorted([int(i.split()[0]) for i in data]), sorted([int(i.split()[1]) for i in data])
    print(sum([abs(a[i] - b[i]) for i in range(len(a))]))

# 2166959

# Part 2
def make_dict(data):
    c = {}
    for i in data:
        if i not in c:
            c[i] = 1
        else:
            c[i] += 1
    return c
with open("D1.txt", "r") as f:
    data = f.read().splitlines()
    data_0 = [int(i.split()[0]) for i in data]
    data_1 = [int(i.split()[1]) for i in data]
    a = make_dict(data_0)
    b = make_dict(data_1)
    # print(a)
    # print(b)
    print(sum([a[i] * i * b[i] for i in a.keys() if i in b.keys()]))

# 23741109

File: Day_1/Historian_Hysteria.go
package main

import (
	"fmt"
    "os"
	"strings"
	"slices"
	"math"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

var print = fmt.Println

func makeDict(data []int) map[int]int {
    c := make(map[int]int)
    for _, item := range data {
        c[item]++
    }
    return c
}


func main() {
	// Part 1
	dat, err := os.ReadFile("D1.txt")
	check(err)
	var data string = string(dat)
	var lines = strings.Split(strings.ReplaceAll(data, "\r\n", "\n"), "\n")
	a := make([]int, len(lines))
	b := make([]int, len(lines))
	for i, line := range lines {
		fmt.Sscanf(strings.Split(line,"   ")[0], "%d", &a[i])
		fmt.Sscanf(strings.Split(line,"   ")[1], "%d", &b[i])
	}
	slices.Sort(a)
	slices.Sort(b)
	var sum int =0
	for i:=0; i<len(a); i++{
		sum += int(math.Abs(float64(a[i]-b[i])))
	}
	print(sum)

	// Part 2
	d1 := makeDict(a)
	d2 := makeDict(b)
	var sum2 int =0
	for i:= range d1{
		if val, exist := d2[i]; exist{
			sum2 += int(val*i*d1[i])
		}
	}
	print(sum2)
}